### ISSUE TITLE
Conflict Between Exploding Dice and Keep High/Keep Low Commands

### ISSUE THREAD

**Issue by GreenKnight1294** [dice-roller/rpg-dice-roller#263]
**Description of Bug**
I use a ttrpg system that requires the use of both advantage (KH and KL commands) and exploding dice at the same time.
Currently the way the bot handles this interaction is that it explodes the dice _before_ applying the keep high or keep low commands.

Truth be told I don't know if this could classify as a bug as I haven't seen this interaction used in any other game system other than Open Legend, but the way this interaction should work _as desired by the user_ is that the dice bot should roll all the dice, only then keep the highest or lowest specified, and _then_ explode the resulting dice without applying KH or KL to the additional dice.

**To Reproduce**
Easiest ways to reproduce the behavior:
1. Request the bot a string that produces ample number of low sided dice, such as (roll 8d4kh4!)
2. Note that the results contain one or more 4's, which will make the dice roller add additional d4 rolls to the result
3. Note how from the additional rolled dice only the 4 highest will be selected as results
4. The above means that the dice roller will never produce results higher than 16, which it should if it explodes enough dice.

**Visual Example and Expected Behavior**
To show the intended way for this interaction to resolve I'll use the dice bot I was utilizing with this library
![image](https://user-images.githubusercontent.com/120273637/206869449-fa74429b-43b7-4ee3-9bed-19a6dc1ecac7.png)
As can be observed the first 8 rolls obtained were 3, 4, 2, 2, 3, 2, 3, 3. Because one of these rolls was a 4, the dice roller adds another d4 at the end, which rolls a 3.
The dice bot then proceeds to remove 5 dice from the equation and only keep the 4 and 3's shown.
-The intended behavior should have been to keep the highest 4 numbers first (4, 3, 3, 3) and only then add another d4 for the result of 4 (3). In this case the end result should have been 16 and not 13.

**Environment**
 - I'm using the current version of the bot Dice Witch and was referred to this github

Thank you for your time. If you'd like to reach out over Discord feel free to request my user.

**Comment by GreenImp** [comment]
Hey, thanks for the ticket. Nice to hear about another system using my dice roller! I've not come across Dice Witch before.

You've raised a very interesting point about the execution order of the modifiers. As [mentioned in the docs](https://dice-roller.github.io/documentation/guide/notation/modifiers.html), modifiers are explicitly run in a particular order.

The current order was intentional, but also originally done with the intention of making it feature compatible with how Rol20 does things. With that, it uses the following logic (Found on the [Roll20 documentation](https://help.roll20.net/hc/en-us/articles/360037773133-Dice-Reference#DiceReference-SortingDice(B,F)sa/sd)):

> Modifiers are applied in the following order:
>
> * Exploding, Compounding, Penetrating, Rerolls
> * These are applied while the dice are still rolling. imagine that you roll some exploding 6d6 on your table and the instant one of them settles as a 6 another die is rolled
> * Keep, Drop, Success, Failure, Sorting
> * These are applied after all of the dice have "settled" and all of the result values are known.

I _think_ that the order here makes sense, as we roll all of the possible dice, before we look at modifying the rolls.

I get what you're saying though, and I'm honestly not sure whether your suggestion is preferable or not.

I guess that part of it depends on whether, when rolling something like a "keep highest 4", you would ever expect to receive more than 4 rolls. Which I don't think I would.
If I roll `8d4kh4`, it would return 4 rolls. If I roll `8d4!kh4`, I think I'd expect it to still only ever return 4 rolls, even if some of them exploded. Whereas you're saying that you could end up with more, if any dice exploded.

Do you have a link to the documentation on Open Legend that specifies this behaviour? I had a brief look, but couldn't find anything.

I think this is a difficult one to sort out. I'm happy to discuss this, and get some more opinions (Ideally from a range of different games). might even be worth moving this to a discussion.

**Comment by GreenKnight1294** [comment]
Interestingly the way Roll20 does it is also incompatible with Open Legend. I'm glad there was some logic behind the order for sure. And it makes sense because Roll20, as far as I know, has not changed the way their dice roller works for years, whereas OL came out around 3 years ago. The OL community actually has their own dice roller, but I like to experiment with different engines when I can.

I have no clue if there's any other system that currently uses exploding dice interacting with KH and KL, I doubt there is. Below is the documentation on Open Legend and the way it rolls the dice:
![image](https://user-images.githubusercontent.com/120273637/207000165-9b628516-47a6-4ec3-b721-50175508dba8.png)
[Source: https://openlegendrpg.com/core-rules/actions-attributes]

In Open Legend exploding dice adding additional numbers to an attack is an integral part of how "critical hits" work, and there is no limit to how high of a result you can get, which is why you end up wanting more than the original kh/kl amount.

**Comment by GreenKnight1294** [comment]
Also you don't roll 8d8's, that was just to show an easily reproducible example. In the system you tend to roll 1d20 and add modifier dice such as 2d6. Advantage in the system adds to the number of dice rolled to the modifier dice (ie: advantage 3 = 5d6kh2 and disadvantage = kl) + 1d20. The idea is that advantage will have you roll higher base numbers, and only then explode them as normal.

It makes more sense the way I'm describing because when you _don't_ use advantage you can keep exploding dice infinitely. Whereas if you use advantage with the KH/KL command it automatically caps the amount of dice rolled to the KH/KL number.

**Comment by GreenKnight1294** [comment]
Would it take too much work to change the order in which the commands are applied?

**Comment by GreenImp** [comment]
Sorry, it's unlikely to be changed as it would affect anyone using it with the current behaviour.

You're welcome to fork the repo with the change in order - it's quite a simple change. Each modifier has an `order` property, set in their constructor:

https://github.com/dice-roller/rpg-dice-roller/blob/d6318e2e8f51d57353b0bb696410f1dd0daf0650/src/modifiers/ExplodeModifier.js#L32

You would need to change it so that the explode modifier order is greater than the keep / drop modifier orders.
Probably to the following:

```javascript
DropModifer.order = 3
KeepModifier.order = 4
ExplodeModifier.order = 5
```

**Comment by GreenImp** [comment]
Just a thought actually; it might be possible to make this customisable using static class properties for the order instead.

Something like:

```javascript
ExplodeModifier extends ComparisonModifier {
  static order = 3;
}
```

And then, to change it, you should be able to do something like:

```javascript
import { DropModifer, ExplodeModifier, KeepModifier } from '@dice-roller/rpg-dice-roller';

DropModifer = 3;
KeepModifier = 4;
ExplodeModifier.order = 5;
```

But that'll require some code re-working and checking to make sure it works.

I'm open to PRs.

**Comment by GreenKnight1294** [comment]
Is there a TTRPG system where the current order of commands is used? It doesn't make sense to me that you would want to explode dice only to keep the highest values from the exploded pool of dice and discard the rest.

It almost feels more like an oversight that keeps being left over in different dice engines that do not expect people to need this.

**Comment by GreenImp** [comment]
> Is there a TTRPG system where the current order of commands is used?

Honestly, I don't know. Personally, I can see the reason for it working both ways, but I can't really change the functionality because I have no way of knowing how many people use and expect the current behaviour.

> Also, when you're suggesting I fork the bot, you're talking about hosting my own instance of the dice bot with this change enabled? Would it be possible for me to use it on a pre-existing bot like Dice Witch?

No, you're correct, forking the repo won't solve the situation with how Dice Witch functions, unless you also host your own version of Dice Witch, which obviously isn't ideal.

I think that, at present your best option is to use the workaround [suggested above](https://github.com/dice-roller/rpg-dice-roller/issues/263#issuecomment-1503179356):

```javascript
3d8!!kh1
```

If it were possible to change the modifier execution order, as [mentioned above](https://github.com/dice-roller/rpg-dice-roller/issues/263#issuecomment-1503084670), it would allow third-party libraries to change the execution order themselves.
Dice Witch could then initialise the library with that order by default, or perhaps add an argument to the bot, to allow users to change it. Obviously that would require a change on Dice Witch as well.

**Comment by GreenKnight1294** [comment]
Thank you. The code you provided as an alternative seems to work surprisingly fine.

**Comment by GreenImp** [comment]
That's great! I'm going to close this ticket, but I've raised a separate one for looking into enabling changing of the modifier execution order; #279

**Comment by GreenImp** [comment]
@GreenKnight1294 as an update, I've added in the functionality to change the modifier execution orders, which has been released in v5.4.0. See #282 

The docs have been updated to explain the process: https://dice-roller.github.io/documentation/guide/customisation.html#modifier-execution-order

The [example in the docs](https://dice-roller.github.io/documentation/guide/customisation.html#modifier-execution-order) shows the necessary order for Open Legends.

This means that anyone implementing this library could build a way of changing it.

For example, Dice Witch could potentially read a parameter passed through, and then modify the orders based on that:

```javascript
/roll notation:8d4kh4! openlegend:true
```

```javascript
import { DiceRolller, Modifiers } from '@dice-roller/rpg-dice-roller';

const setOpenLegendOrder = () => {
  Modifiers.KeepModifier.order = 3;
  Modifiers.DropModifier.order = 4;
  Modifiers.ExplodeModifier.order = 5;
  Modifiers.ReRollModifier.order = 6;
  Modifiers.UniqueModifier.order = 7;
};

// check the parameter value here
if (openlegend) {
  setOpenLegendOrder();
}

const roll = new DiceRoll('8d4kh4!');
```

And, if necessary, it could be reset back to the defaults with something like:

```javascript
const resetOrder = () => {
  Modifiers.KeepModifier.order = 6;
  Modifiers.DropModifier.order = 7;
  Modifiers.ExplodeModifier.order = 3;
  Modifiers.ReRollModifier.order = 4;
  Modifiers.UniqueModifier.order = 5;
};
```

Obviously this is just a very rough example.
I'd suggest raising it on the [Dice Witch GitHub repo](https://github.com/cnharrison/dice-witch), and see what they suggest.
